aasx.adapter: fix semantic id type deserialization of `ModelReference`

JGrothoff opened the pull request:
Many AASX packages include semantic ids with external or model references.

## Concept descriptions are lost when reading and writing back to AASX

### Problem

When a semantic id with a `ModelReference` is deserialized its `type` is currently set to the abstract type [`Referable`](https://github.com/eclipse-basyx/basyx-python-sdk/blob/c360897422c6cf483af59b090d2c6e1e1c9507e7/sdk/basyx/aas/adapter/json/json_deserialization.py#L321).
When the aasx is serialized again and the semantic id refers to a concept description, it is [skipped, because only concrete `ModelReference` types](https://github.com/eclipse-basyx/basyx-python-sdk/blob/c360897422c6cf483af59b090d2c6e1e1c9507e7/sdk/basyx/aas/adapter/aasx.py#L403) are looked up and serialized.

### Solutions

* Suggested: Set the `type` of the semantic id to the concrete type determined by the last key element.
* Alternative: Skip the type check as errors [are caught anyway](https://github.com/eclipse-basyx/basyx-python-sdk/blob/c360897422c6cf483af59b090d2c6e1e1c9507e7/sdk/basyx/aas/adapter/aasx.py#L412) (EAFP style). Maybe lowering the loglevel to warning or info.


## Writing aasx with external references as semantic ids produce many info logs

### Problem

Often an external reference with a global id (e.g. to ECLASS IRDIs) is used as semantic id.
A typical technical data submodel may easily contain around 100 of them.
Writing to an aasx will [print an info message](https://github.com/eclipse-basyx/basyx-python-sdk/blob/c360897422c6cf483af59b090d2c6e1e1c9507e7/sdk/basyx/aas/adapter/aasx.py#L405) for each semantic id, that is not a model reference to an concept description, which clutters the output.

### Solutions

* Suggested: skip external references without log message
* Alternative: lower logging level to debug or delete log info.

JGrothoff commented:
Thanks for investigating the topic so far!
I will try to add or update the tests to cover the issue (thats why i marked this as Draft).

JGrothoff commented:
I added a test that covers all semantic ids that are expected to be ModelReferences for the same type.

s-heppner commented:
This is what we concluded in our meeting: 

- We cannot allow writing wrong `AASX`s. If the `ConceptDescription` of the `ModelReference` is not found, we should raise an Error as the `AASX` is wrong. 
- However, we should obviously not silently discard content, as in not warning the user about the issue. 
- We should try to declutter our output (if there's not actually an issue with the `ExternalReference`). This message may be now unnecessary and might have been historically grown

Sadly, I did not have time yet to look deeper into the matter.

JGrothoff commented:
> * We cannot allow writing wrong `AASX`s. If the `ConceptDescription` of the `ModelReference` is not found, we should raise an Error as the `AASX` is wrong.

I rebased on main and changed to a re-raise for those cases. It will raise an error if the reference is a `ModelReference` and has the type `ConceptDescription`, but it can't be resolved in the given `object_store`.
 Of couse, we could change the Exeception class here, but i will leave that up to you.
 
 This will still not check if
 * an abstract `Reference` or
 * a `ModelReference` with `type` other than a `ConceptDescription`
 can be resolved and is contained in the `object_store` (or in the `objects_to_be_written`).
 But it will log an info message about that, like before.

s-heppner commented on sdk/basyx/aas/examples/data/_helper.py:
```diff
@@ -212,7 +212,15 @@ def _check_has_semantics_equal(self, object_: model.HasSemantics, expected_objec
         :return: The value of expression to be used in control statements
         """
-        self.check_attribute_equal(object_, "semantic_id", expected_object.semantic_id)
+        if self.check_attribute_equal(object_, "semantic_id", expected_object.semantic_id) and \
+                isinstance(expected_object.semantic_id, model.ModelReference):
```
This if statement only checks `semantic_id`s that are `ModelReference`s . We should also check for string equality of the `semantic_id` if they aren't a `ModelReference`.

Note that there's inconclusive discussions in the IDTA atm, as to when two semantic IDs are to be considered "equal", for example, IRDIs can be written in ECLASS notation: `0173-1#01-ACE548#009` or IEC 61360 CDD notation: `0112/2///61360_4#AAA012`. I wouldn't want to start implementing any `semantic_id` equalivalence here other than string equivalence until there's specification telling us how to do it, otherwise we'll loose interoperability.

JGrothoff replied:
I think python evaluates boolean conditions lazily, thus the `check_attribute_equal` testing the string equality of the semantic ids will be executed as before. But i agree that it's not obvious and i will move it before the if statement, which is prettier anyway.

s-heppner commented on sdk/basyx/aas/adapter/json/json_deserialization.py:
```diff
@@ -340,10 +340,12 @@ def _construct_model_reference(cls, dct: Dict[str, object], type_: Type[T], obje
                            keys[-1].type.name, " / ".join(str(k) for k in keys), type_.__name__)
-        return object_class(tuple(keys), type_, cls._construct_reference(_get_ts(dct, 'referredSemanticId', dict))
+        return object_class(tuple(keys), last_key_type,
+                            cls._construct_reference(_get_ts(dct, 'referredSemanticId', dict))
```
I don't know enough about the rules for `Reference` objects to determine this, either. 
Could you add a few lines of comments of what happens here, so that someone can fix it, if it ends up breaking anything?

s-heppner commented on sdk/basyx/aas/adapter/aasx.py:
```diff
@@ -402,19 +402,22 @@ def write_aas(self,
             for semantic_id in traversal.walk_semantic_ids_recursive(identifiable):
+                if isinstance(semantic_id, model.ExternalReference):
+                    continue
                 if not isinstance(semantic_id, model.ModelReference) \
                         or semantic_id.type is not model.ConceptDescription:
                     logger.info("semanticId %s does not reference a ConceptDescription.", str(semantic_id))
```
I'm not entirely sure while this was written the way it was. I would expect that most `semantic_id`s do not refer to a `ConceptDescription`. 
I would remove the logger.info statement completeley.

JGrothoff replied:
I agree. We should skip `ExternalReference` anyway, as they are probably not included in the `object_store`. I will just remove the log.

s-heppner commented on sdk/basyx/aas/adapter/aasx.py:
```diff
@@ -402,19 +402,22 @@ def write_aas(self,
-                    logger.error("semanticId %s resolves to %s, which is not a ConceptDescription",
-                                 str(semantic_id), e.value)
-                    continue
+                    raise model.UnexpectedTypeError(
+                        e.value,
+                        "semanticId {} resolves to {}, which is not a ConceptDescription".format(
+                            str(semantic_id), e.value))
```
Are we certain that it is not allowed for a `semantic_id` to be a `ModelReference` but does not refer to `ConceptDescription`, disregarding the sensibility of such a `semantic_id`? Is there a constraint in the specification that forbids this?

JGrothoff replied:
I just looked at the rest of the code and the functions doc string, stating that it [will skip wrong ConceptDescription references](https://github.com/eclipse-basyx/basyx-python-sdk/blob/a441d203a11ce2d009e032c778501bec7d0652b2/sdk/basyx/aas/adapter/aasx.py#L369) and even missing Submodels.

Thus i change my mind and prefere the logging as before, but with warning instead of info level on `KeyError`. I think this is more practical, as i would rather have a list of warnings when writing the aasx, then stopping on the first missing submodel or concept description.
Of couse, it would be nice to have a "strict mode", where it raises directly, to catch the exceptions programmatically.
